src/TenantSocketConnection.js

```diff
@@ -0,0 +1,223 @@
+/* @flow */
+
+import {
+  parse,
+  subscribe,
+  type ExecutionResult,
+  type GraphQLSchema,
+} from 'graphql';
+import IoServer from 'socket.io';
+
+import * as asyncUtils from './asyncUtils';
+import type RedisSubscriber from './RedisSubscriber';
+
+type HttpApi = {
+  setToken(token: ?string): void,
+};
+
+export type Credentials = {
+  expSeconds: number,
+  tenants: { [tenantId: string]: number },
+};
+
+export type Context = {
+  httpApi: HttpApi,
+};
+
+type Subscription = {
+  id: string,
+  query: string,
+  variables: Object,
+};
+
+type MaybeSubscription = Promise<
+  AsyncIterator<ExecutionResult> | ExecutionResult,
+>;
+
+type ServerConfig = {|
+  socket: IoServer.socket,
+  redis: RedisSubscriber,
+  context: { httpApi: HttpApi },
+  schema: GraphQLSchema,
+  hasPermission: (data: any, socketCredentials: Credentials) => boolean,
+  defaultParseMessage: (data: string) => any,
+  maxSubscriptionsPerConnection?: number,
+  gracePeriodSeconds: number,
+  fetchCredentials: (
+    context: Context,
+    authorization: string,
+  ) => Promise<Credentials>,
+|};
+
+const SECONDS_TO_MS = 1000;
+
+const acknowledge = cb => {
+  if (cb) cb();
+};
+
+function isExpired(expSeconds: ?number): boolean {
+  return expSeconds != null && !!(Date.now() > expSeconds * SECONDS_TO_MS);
+}
+
+// TenantSocketConnection manages a single socket connection for a single user.
+// Includes
+//  - authorization,
+//  - authentication,
+//  - and some rudimentary connection constraints (max connections).
+export default class TenantSocketConnection {
+  config: ServerConfig;
+
+  subscriptions: Map<string, MaybeSubscription>;
+  credentials: {
+    expSeconds: number,
+    tenants: { [tenantId: string]: number },
+  };
+  renewTimer: ?number;
+
+  constructor(config: ServerConfig) {
+    this.config = config;
+    this.credentials = { expSeconds: 0, tenants: {} };
+    this.subscriptions = new Map();
+    this.config.socket
+      .on('authenticate', this.handleAuthenticate)
+      .on('subscribe', this.handleSubscribe)
+      .on('unsubscribe', this.handleUnsubscribe)
+      .on('disconnect', this.handleDisconnect);
+  }
+
+  isAuthenticated = () =>
+    this.credentials.tenants && !isExpired(this.credentials.expSeconds);
+
+  isAuthorized = (data: any) =>
+    !!(
+      this.isAuthenticated() &&
+      this.config.hasPermission(data, this.credentials)
+    );
+
+  scheduleRenew(implicitToken: string) {
+    const deltaMS = this.credentials.expSeconds * SECONDS_TO_MS - Date.now();
+    const deltaMSAdjusted = Math.max(
+      0,
+      deltaMS - this.config.gracePeriodSeconds * SECONDS_TO_MS,
+    );
+    this.renewTimer = setTimeout(() => {
+      this.handleAuthenticate(implicitToken);
+    }, deltaMSAdjusted);
+  }
+
+  async updateCredentials(implicitToken: string) {
+    this.credentials = await this.config.fetchCredentials(
+      this.config.context,
+      implicitToken,
+    );
```

Comment:
@taion is your thought we should eventually split most of this stuff out into a CredentialManager?

Reply:
yeah, it should look like:

```js
type CredentialsManager<Credentials> {
  updateAuthorization: (authorization: string) => void,
  getCredentials: () => Credentials,
  close: () => void, // Or maybe some async thing.
};
```

And you pass in `context` when you construct it via a:

```js
createCredentialsManager: (context: Context) => CredentialsManager<Credentials>,
```